mesh/nurbs.cpp

```diff
@@ -2330,14 +2363,78 @@ NURBSExtension::NURBSExtension(Mesh *mesh_array[], int num_pieces)
    MergeWeights(mesh_array, num_pieces);
 }
 
-NURBSExtension::~NURBSExtension()
+NURBSExtension::NURBSExtension(const Mesh *patch_topology,
+                               const Array<const NURBSPatch*> p)
 {
-   if (patches.Size() == 0)
+   patchTopo = new Mesh( *patch_topology );
+   patchTopo->GetEdgeVertexTable();
+   own_topo = 1;
+   patches.Reserve(p.Size());
+   Array<int> edges;
+   Array<int> oedges;
+   Array<int> kvs(3);
+   edge_to_knot.SetSize(patch_topology->GetNEdges());
+   NumOfKnotVectors = 0;
+   NumOfElements = 0;
+   for (int ielem = 0; ielem < patch_topology->GetNE(); ++ielem)
    {
-      delete bel_dof;
-      delete el_dof;
+      patches.Append(new NURBSPatch(*p[ielem]));
+      NURBSPatch& patch = *patches[ielem];
+      int num_patch_elems = 1;
+      for (int ikv = 0; ikv < patch.GetNKV(); ++ikv)
+      {
+         kvs[ikv] = knotVectors.Size();
+         knotVectors.Append(new KnotVector(*patch.GetKV(ikv)));
+         num_patch_elems *= patch.GetKV(ikv)->GetNE();
+         ++NumOfKnotVectors;
+      }
+      NumOfElements += num_patch_elems;
+      patch_topology->GetElementEdges(ielem, edges, oedges);
+      for (int iedge = 0; iedge < edges.Size(); ++iedge)
+      {
+         if (iedge < 8)
+         {
+            if (!(iedge & 1))
+            {
+               edge_to_knot[edges[iedge]] = kvs[0];
+            }
+            else
+            {
+               edge_to_knot[edges[iedge]] = kvs[1];
+            }
+         }
+         else
+         {
+            edge_to_knot[edges[iedge]] = kvs[2];
+         }
+      }
    }
 
+   GenerateOffsets();
+   CountBdrElements();
+   NumOfActiveElems = NumOfElements;
+   activeElem.SetSize(NumOfElements);
+   activeElem = true;
+
+   CreateComprehensiveKV();
+   SetOrdersFromKnotVectors();
+
+   GenerateActiveVertices();
+   InitDofMap();
+   GenerateElementDofTable();
+   GenerateActiveBdrElems();
+   GenerateBdrElementDofTable();
+
+   weights.SetSize(GetNDof());
+
+   CheckPatches();
+}
+
+NURBSExtension::~NURBSExtension()
+{
+   if (bel_dof) { delete bel_dof; }
```

Comment:
Should these be initialized to `nullptr` by all constructors?

Reply:
I have the default constructor initialize these to null, and double-checked to verify that all other `NURBSExtension` constructors call `GenerateElementDofTable` and `GenerateBdrElementDofTable`, which set these to non-null values, so I don't think there's any risk of freeing an invalid pointer here.